rust/src/treefile.rs

```diff
@@ -576,10 +576,42 @@ impl TreefileExternals {
 /// array elements.
 fn whitespace_split_packages(pkgs: &[String]) -> Vec<String> {
     pkgs.iter()
-        .flat_map(|pkg| pkg.split_whitespace().map(String::from))
+        .flat_map(|element| split_whitespace_unless_quoted(element).map(String::from))
         .collect()
 }
 
+// Helper for whitespace_split_packages().
+// Splits a String by whitespace unless substring is wrapped between single quotes
+// and returns split Strings in an Iterator, with quoted packages first.
+fn split_whitespace_unless_quoted(element: &String) -> impl Iterator<Item = String> {
+    let mut quoted_pkgs: Vec<String> = vec![];
+    let mut to_whitespace_split: Vec<String> = vec![];
+    let mut start_index = 0;
+    let mut looping_over_quoted_pkg = false;
+    for (i, c) in element.chars().enumerate() {
+        if c == '\'' {
+            if looping_over_quoted_pkg {
+                quoted_pkgs.push(String::from(&element[start_index..i]));
+                looping_over_quoted_pkg = false;
+            } else {
+                to_whitespace_split.push(String::from(&element[start_index..i]));
+                looping_over_quoted_pkg = true;
+            }
+            start_index = i + 1;
+        }
+        if i == element.len() - 1 {
+            to_whitespace_split.push(String::from(&element[start_index..]));
+        }
+    }
+    let mut ret: Vec<String> = vec![];
+    ret.extend(quoted_pkgs);
```

Comment:
Let's move this to the top - i.e. directly push into `ret` rather than having a separate vector.

Reply:
Right, this would make a lot more sense.